enhancements/api-review/criocredentialproviderconfig-for-namespace-scoped-mirror-authentication.md

```diff
@@ -0,0 +1,561 @@
+
+- Container Runtime Config Controller watches for `CRIOCredentialProviderConfig` changes and updates `CredentialProviderConfig` file `/etc/kubernetes/credential-providers/<platform>-credential-provider.yaml` with `crio-credential-provider` configuration to all nodes
+
+- Machine Config Operator restarts kubelet to make the new CredentialProviderConfig settings effective
+
+```yaml
+# example /etc/kubernetes/credential-providers/ecr-credential-provider.yaml
+apiVersion: kubelet.config.k8s.io/v1
+kind: CredentialProviderConfig
+providers:
+  - name: crio-credential-provider
+    matchImages:
+      - "docker.io"
+    defaultCacheDuration: "1s"
+    apiVersion: credentialprovider.kubelet.k8s.io/v1
+    tokenAttributes:
+      serviceAccountTokenAudience: https://kubernetes.default.svc
+      cacheType: "Token"
+      requireServiceAccount: false
+  # existing ecr-credential-provider config on nodes
+  - name: ecr-credential-provider
+    matchImages:
+      - "*.dkr.ecr.*.amazonaws.com"
+      - "*.dkr.ecr.*.amazonaws.com.cn"
+      - "*.dkr.ecr-fips.*.amazonaws.com"
+      - "*.dkr.ecr.us-iso-east-1.c2s.ic.gov"
+      - "*.dkr.ecr.us-isob-east-1.sc2s.sgov.gov"
+    defaultCacheDuration: "12h"
+    apiVersion: credentialprovider.kubelet.k8s.io/v1
+```
+
+##### Step 2: Application Deployment
+
+1. Application Developer creates Pod spec referencing images from source registries (docker.io/nginx)
+
+```yaml
+# app-pod.yaml
+apiVersion: v1
+kind: Pod
+metadata:
+  name: web-app
+  namespace: app-team-alpha
+spec:
+  containers:
+  - name: web-server
+    image: docker.io/nginx@sha256:abcdef... # actual image pull will be from mirror.example.com/nginx@sha256:abcdef...
+```
+
+##### Step 3: Image Pull Process
+
+1. Kubelet processes Pod creation request and identifies that the image  `docker.io/nginx@sha256:abcdef...` matches against `crio-credential-provider` `matchImages` in the `CredentialProviderConfig` file.
+
+2. Kubelet executes `crio-credential-provider` binary, requests for service account token for Pod and sends `CredentialProviderRequest` to the binary with Pod context (namespace, service account) through stdin.
+
+3. `crio-credential-provider` execution:
+
+ - Parses the `CredentialProviderRequest` from stdin and gets Pod namespace(app-team-alpha) from it
+
+ - resolves mirrors from `/etc/containers/registries.conf` for source image repository(docker.io/nginx)
+
+ - Finds mirror pull secrets in Pod namespace (team-alpha-mirror-secret) by using the service account token from the request
+
+ - Generates a short-lived authentication file for each image pull at `/etc/crio/auth/<NAMESPACE>-<IMAGE_NAME_SHA256>.json`, which includes mirror credentials, source registry credentials, and any global pull secrets.
+
+ - Returns empty `CredentialProviderResponse` to kubelet
+
+4. CRI-O image pull:
+
+  - Detects the file `/etc/crio/auth/<NAMESPACE>-<IMAGE_NAME_SHA256>.json` exists, uses the auth file for image pull authentication
```

Comment:
In previous 'Component Changes & Interactions' part, it describes: 
If the auth file generated by `crio-credential-provider` exists, moves the auth file from `/etc/crio/auth/<NAMESPACE>-<IMAGE_NAME_SHA256>.json` to a unique temporary location (`/etc/crio/auth/in-use/<NAMESPACE>-<IMAGE_NAME_SHA256>-<UUID>.json`) for each image pull

So it's better to add the part of moving auth file to a unique temporary directory